src/AutoRest.CSharp/Common/AutoRest/Plugins/GeneratedCodeWorkspace.cs

```diff
@@ -170,5 +171,236 @@ public async Task<CSharpCompilation> GetCompilationAsync()
+            var visited = new HashSet<SyntaxNode>();
+            while (queue.Count > 0)
+            {
+                var node = queue.Dequeue();
+                if (visited.Contains(node))
+                    continue;
+                visited.Add(node);
+                // add this node to the public list
+                yield return node;
+                // add every type referenced by this node to the queue
+                await foreach (var child in GetReferencedTypes(compilation, node))
+                {
+                    queue.Enqueue(child);
+                }
+            }
+        }
+
+        private async IAsyncEnumerable<SyntaxNode> GetReferencedTypes(Compilation compilation, SyntaxNode root)
+        {
+            var semanticModel = compilation.GetSemanticModel(root.SyntaxTree);
+            var publicMemberVisitor = new PublicMemberVisitor();
+            publicMemberVisitor.Visit(root);
+            foreach (var member in publicMemberVisitor.PublicMembers)
+            {
+                var symbol = semanticModel.GetDeclaredSymbol(member);
+                if (symbol == null)
+                    continue;
+                var list = new HashSet<SyntaxNode>();
+                await ProcessSymbol(symbol, list);
+                foreach (var node in list)
+                {
+                    yield return node;
+                }
+            }
+        }
+
+        private async Task ProcessSymbol(ISymbol? symbol, HashSet<SyntaxNode> result)
+        {
+            if (symbol == null || symbol.DeclaredAccessibility != Accessibility.Public)
+                return;
+            switch (symbol)
+            {
+                case INamedTypeSymbol typeSymbol:
+                    foreach (var reference in typeSymbol.DeclaringSyntaxReferences)
+                    {
+                        result.Add(await reference.GetSyntaxAsync());
+                    }
+                    await ProcessSymbol(typeSymbol.BaseType, result);
+                    foreach (var typeArgument in typeSymbol.TypeArguments)
+                    {
+                        await ProcessSymbol(typeArgument, result);
+                    }
+                    break;
+                case IMethodSymbol methodSymbol:
+                    await ProcessSymbol(methodSymbol.ReturnType, result);
+                    foreach (var parameter in methodSymbol.Parameters)
+                    {
+                        await ProcessSymbol(parameter.Type, result);
+                    }
+                    break;
+                case IPropertySymbol propertySymbol:
+                    await ProcessSymbol(propertySymbol.Type, result);
+                    break;
+                default:
+                    throw new InvalidOperationException($"Not implemented for symbol {symbol.GetType()}");
+            }
+        }
+
+        public async Task RemoveOrphanedModels(ImmutableHashSet<string> modelsToKeep)
```

Comment:
Rename the method to something like `InternalizeOrphanedModels`?

Reply:
Make sense since we are not removing anything